Pillar based configuration template

tacerus opened the pull request:
Allow management of general configuration using pillar data.

<!--
Please fill in this PR template to make it easier to review and merge.

It has been designed so that a lot of it can be completed *after* submitting it,
e.g. filling in the checklists.

Notes:
1. Please keep the PR as small as is practicable; the larger a PR gets, the harder it becomes to review and the more time it requires to get it merged.
2. Similarly, please avoid PRs that cover more than one type; it should be a _bug fix_ *OR* a _new feature_ *OR* a _refactor_, etc.
3. Please direct questions to the [`#formulas` channel on Slack](https://saltstackcommunity.slack.com/messages/C7LG8SV54/), which is bridged to `#saltstack-formulas` on Freenode.
4. Feel free to suggest improvements to this template by reporting an issue or submitting a PR.  The source of this template is:
  - https://github.com/saltstack-formulas/.github/blob/master/.github/pull_request_template.md
-->

### PR progress checklist (to be filled in by reviewers)
<!-- Please leave this checklist for reviewers to tick as they work through the PR. -->

- [x] Changes to documentation are appropriate (or tick if not required)
- [x] Changes to tests are appropriate (or tick if not required)
- [ ] Reviews completed

---

### What type of PR is this?
<!-- Please tick each box that is relevant (after creating the PR). -->

Allow general NSD configuration management using pillar data.

#### Primary type
<!-- There really should be only *one* of these types ticked for each PR. -->

- [ ] `[build]`    Changes related to the build system
- [ ] `[chore]`    Changes to the build process or auxiliary tools and libraries such as documentation generation
- [ ] `[ci]`       Changes to the continuous integration configuration
- [x] `[feat]`     A new feature
- [ ] `[fix]`      A bug fix
- [ ] `[perf]`     A code change that improves performance
- [ ] `[refactor]` A code change that neither fixes a bug nor adds a feature
- [ ] `[revert]`   A change used to revert a previous commit
- [ ] `[style]`    Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc.)

#### Secondary type
<!-- Most PRs should include all of the following types as well. -->

- [ ] `[docs]`     Documentation changes
- [x] `[test]`     Adding missing or correcting existing tests

### Does this PR introduce a `BREAKING CHANGE`?
<!-- If so, change the following to a `Yes` and explain what the breaking changes are. -->
<!-- If there are multiple breaking changes, list them all. -->

No.

### Related issues and/or pull requests
<!-- Please link any related issues/PRs here, especially any issues that are closed by this PR. -->



### Describe the changes you're proposing
<!-- A clear and concise description of what you have implemented. -->
<!-- Consider explaining each commit if they cover different aspects of the proposed changes. -->

- Generate `10-salt.conf` using the `nsd:config_data` pillar.
- Allow zones to be added with custom options and source files

### Pillar / config required to test the proposed changes
<!-- Provide links to the SLS files and/or relevant configs (be sure to remove sensitive info). -->

```
nsd:
  config_data:
    server:
      hide-version: 'yes'
      verbosity: 2
    remote-control:
      control-enable: 'yes'
      control-interface: /run/nsd.sock
  zones:
    example.com:
      zone_source: salt://zones/test.zone
      request-xfr: '127.0.0.1@5353'
```


### Debug log showing how the proposed changes work
<!-- Include a debug log showing how these changes work, e.g. using `salt-minion -l debug`. -->
<!-- Alternatively, linking to Kitchen debug logs is useful, e.g. via. Travis CI. -->
<!-- Most useful is providing a passing InSpec test, which can be used to verify any proposed changes. -->

GitHub does not let me to paste such long output.

### Documentation checklist
<!-- Please tick each box that is relevant (after creating the PR). -->

- [ ] Updated the `README` (e.g. `Available states`).
- [x] Updated `pillar.example`.

### Testing checklist
<!-- Please tick each box that is relevant (after creating the PR). -->

- [ ] Included in Kitchen (i.e. under `state_top`).
- [ ] Covered by new/existing tests (e.g. InSpec, Serverspec, etc.).
- [x] Updated the relevant test pillar.

### Additional context
<!-- Add any other context about the proposed changes here. -->

Do I need to duplicate the test pillar in all operating systems under `test/integration/default/files/_mapdata/`?

tacerus committed:
feat(config): pillar based config template

Allow management of general configuration using pillar data.

Signed-off-by: Georg Pfuetzenreuter <mail@georg-pfuetzenreuter.net>

tacerus commented:
The tests are failing on changes prior to mine.

baby-gnu reviewed (changes requested):
I tried locally and it's failing:

<details><summary><code>./bin/kitchen verify default-opensuse-leap-152-master-py3</code></summary>

```
-----> Starting Test Kitchen (v3.3.1)
-----> Setting up <default-opensuse-leap-152-master-py3>...
       Finished setting up <default-opensuse-leap-152-master-py3> (0m0.00s).
-----> Verifying <default-opensuse-leap-152-master-py3>...
       Loaded default 

Profile:   nsd formula (default)
Version:   (not specified)
Target:    ssh://kitchen@localhost:49155
Target ID: 310008fd-79bc-5145-b8fa-733dc0f45c44

  ✔  nsd.service.running: The service should be installed, enabled and running
     ✔  Service nsd is expected to be installed
     ✔  Service nsd is expected to be enabled
     ✔  Service nsd is expected to be running
     ✔  Command: `nslookup -port=53530 ns1.example.test 127.0.0.1` stdout is expected to match "192.168.0.1"
     ✔  Command: `nslookup -port=53530 ns2.example.test 127.0.0.1` stdout is expected to match "192.168.0.2"
     ✔  Command: `nslookup -port=53530 example.test 127.0.0.1` stdout is expected to match "192.168.0.10"
     ✔  Command: `nslookup -port=53530 www.example.test 127.0.0.1` stdout is expected to match "192.168.0.10"
     ✔  Command: `nslookup -port=53530 mail.example.test 127.0.0.1` stdout is expected to match "192.168.0.11"
     ✔  Command: `nslookup -port=53530 192.168.0.1 127.0.0.1` stdout is expected to match "ns1.example.test"
     ✔  Command: `nslookup -port=53530 192.168.0.2 127.0.0.1` stdout is expected to match "ns2.example.test"
     ✔  Command: `nslookup -port=53530 192.168.0.10 127.0.0.1` stdout is expected to match "example.test"
     ✔  Command: `nslookup -port=53530 192.168.0.11 127.0.0.1` stdout is expected to match "mail.example.test"
  ✔  nsd.package.install: The required package should be installed
     ✔  System Package bash is expected to be installed
  ✔  nsd.config.file: Verify the configuration file
     ✔  File /etc/nsd/nsd.conf is expected to be file
     ✔  File /etc/nsd/nsd.conf is expected to be owned by "root"
     ✔  File /etc/nsd/nsd.conf is expected to be grouped into "root"
     ✔  File /etc/nsd/nsd.conf mode is expected to cmp == "0644"
     ✔  File /etc/nsd/nsd.conf content is expected to include "Your changes will be overwritten."
     ✔  File /etc/nsd/nsd.conf content is expected to include "\"added_in_pillar\": \"pillar_value\""
     ✔  File /etc/nsd/nsd.conf content is expected to include "\"added_in_defaults\": \"defaults_value\""
     ✔  File /etc/nsd/nsd.conf content is expected to include "\"added_in_lookup\": \"lookup_value\""
     ✔  File /etc/nsd/nsd.conf content is expected to include "\"config\": \"/etc/nsd/nsd.conf\""
     ✔  File /etc/nsd/nsd.conf content is expected to include "\"lookup\": {\"added_in_lookup\": \"lookup_value\","
     ✔  File /etc/nsd/nsd.conf content is expected to include "\"pkg\": {\"name\": \""
     ✔  File /etc/nsd/nsd.conf content is expected to include "\"service\": {\"name\": \""
     ✔  File /etc/nsd/nsd.conf content is expected to include "\"tofs\": {\"files_switch\": [\"any/path/can/be/used/here\", \"id\", \"roles\", \"osfinger\", \"os\", \"os_family\"], \"source_files\": {\"nsd-config-file-file-managed\": [\"nsd.conf.jinja\"]}"
     ✔  File /etc/nsd/nsd.conf content is expected to include "\"arch\": \"amd64\""
     ✔  File /etc/nsd/nsd.conf content is expected to include "\"winner\": \"pillar\""
     ✔  File /etc/nsd/nsd.conf content is expected to include "winner of the merge: pillar"
     ✔  File /etc/nsd/nsd.conf.d is expected to be directory
     ✔  File /etc/nsd/nsd.conf.d is expected to be owned by "root"
     ✔  File /etc/nsd/nsd.conf.d is expected to be grouped into "root"
  ×  nsd._mapdata: `map.jinja` should match the reference file
     ×  File content should match profile map data exactly
     
     expected: "---\nvalues:\n  added_in_defaults: defaults_value\n  added_in_lookup: lookup_value\n  added_in_pilla...ontrol-enable: 'yes'\n      control-interface: \"/run/nsd.sock\"\n  zones_dir: \"/etc/nsd/zones\"\n"
          got: "---\nvalues:\n  added_in_defaults: defaults_value\n  added_in_lookup: lookup_value\n  added_in_pilla...ddr.arpa: {}\n    example.test:\n      multi-master-check: false\n  zones_dir: \"/etc/nsd/zones\"\n"
     
     (compared using ==)
     
     Diff:
     @@ -5,6 +5,13 @@
        added_in_pillar: pillar_value
        arch: amd64
        config: "/etc/nsd/nsd.conf"
     +  config_data:
     +    remote-control:
     +      control-enable: 'yes'
     +      control-interface: "/run/nsd.sock"
     +    server:
     +      hide-version: 'yes'
     +      verbosity: 3
        config_files:
          10-server: local-nsd-for-unbound
        config_include_dir: "/etc/nsd/nsd.conf.d"
     @@ -44,12 +51,5 @@
          168.192.in-addr.arpa: {}
          example.test:
            multi-master-check: false
     -  config_data:
     -    server:
     -      hide-version: 'yes'
     -      verbosity: 2
     -    remote-control:
     -      control-enable: 'yes'
     -      control-interface: "/run/nsd.sock"
        zones_dir: "/etc/nsd/zones"



Profile:   InSpec shared resources (share)
Version:   (not specified)
Target:    ssh://kitchen@localhost:49155
Target ID: 310008fd-79bc-5145-b8fa-733dc0f45c44

     No tests executed.

Profile Summary: 3 successful controls, 1 control failure, 0 controls skipped
Test Summary: 32 successful, 1 failure, 0 skipped
>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ActionFailed
>>>>>> Message: 1 actions failed.
>>>>>>     Verify failed on instance <default-opensuse-leap-152-master-py3>.  Please see .kitchen/logs/default-opensuse-leap-152-master-py3.log for more details
>>>>>> ----------------------
>>>>>> Please see .kitchen/logs/kitchen.log for more details
>>>>>> Also try running `kitchen diagnose --all` for configuration
```
</details>

<details><summary><code>./bin/kitchen verify default-debian-10-master-py3</code></summary>

```
[…]
       local:
         Name: dnsutils - Function: pkg.installed - Result: Clean - Started: 10:26:52.293909 - Duration: 45.652 ms
       ----------
                 ID: nsd-mapdata-dump
           Function: file.managed
               Name: /tmp/salt_mapdata_dump.yaml
             Result: True
            Comment: File /tmp/salt_mapdata_dump.yaml updated
            Started: 10:26:52.351076
           Duration: 74.015 ms
            Changes:   
              ----------
              diff:
                  --- 
                  +++ 
                  @@ -7,6 +7,13 @@
                     added_in_pillar: pillar_value
                     arch: amd64
                     config: /etc/nsd/nsd.conf
                  +  config_data:
                  +    remote-control:
                  +      control-enable: 'yes'
                  +      control-interface: /run/nsd.sock
                  +    server:
                  +      hide-version: 'yes'
                  +      verbosity: 3
                     config_files:
                       10-server: local-nsd-for-unbound
                     config_include_dir: /etc/nsd/nsd.conf.d
         Name: nsd - Function: pkg.installed - Result: Clean - Started: 10:26:52.425343 - Duration: 10.949 ms
         Name: /etc/nsd/nsd.conf.d - Function: file.directory - Result: Clean - Started: 10:26:52.436612 - Duration: 1.663 ms
         Name: /etc/nsd/nsd.conf.d/10-server.conf - Function: file.managed - Result: Clean - Started: 10:26:52.439018 - Duration: 71.463 ms
       ----------
                 ID: nsd-config-config_file-file-managed-10-salt
           Function: file.managed
               Name: /etc/nsd/nsd.conf.d/10-salt.conf
             Result: True
            Comment: File /etc/nsd/nsd.conf.d/10-salt.conf updated
            Started: 10:26:52.512140
           Duration: 30.794 ms
            Changes:   
              ----------
              diff:
                  New file
              mode:
                  0644
         Name: /etc/nsd/nsd.conf.d/90-generated-zones.conf - Function: file.managed - Result: Clean - Started: 10:26:52.543700 - Duration: 183.364 ms
       ----------
                 ID: nsd-config-file-file-managed
           Function: file.managed
               Name: /etc/nsd/nsd.conf
             Result: False
            Comment: check_cmd execution failed
              /etc/nsd/nsd.conf.d/10-salt.conf:5: error: duplicate server: element.
              read /home/kitchen/__salt.tmp.xb0jnhpm failed: 1 errors in configuration file
            Started: 10:26:52.727557
           Duration: 26.809 ms
            Changes:   
         Name: /etc/nsd/zones - Function: file.directory - Result: Clean - Started: 10:26:52.754972 - Duration: 1.977 ms
         Name: /etc/nsd/zones/168.192.in-addr.arpa.zone - Function: file.managed - Result: Clean - Started: 10:26:52.757836 - Duration: 15.761 ms
         Name: /etc/nsd/zones/example.test.zone - Function: file.managed - Result: Clean - Started: 10:26:52.773902 - Duration: 23.718 ms
         Name: nsd-control-setup - Function: cmd.run - Result: Clean - Started: 10:26:52.802857 - Duration: 626.576 ms
       ----------
                 ID: nsd-service-running-service-running
           Function: service.running
               Name: nsd
             Result: False
            Comment: Job for nsd.service failed because the control process exited with error code.
              See "systemctl status nsd.service" and "journalctl -xe" for details.
            Started: 10:26:53.475981
           Duration: 122.063 ms
            Changes:   
       ----------
                 ID: nsd-service-control-reload-zones
           Function: cmd.run
               Name: nsd-control reload
             Result: False
            Comment: One or more requisite failed: nsd.service.running.nsd-service-running-service-running
            Started: 10:26:53.599859
           Duration: 0.006 ms
            Changes:   
       
       Summary for local
       -------------
       Succeeded: 11 (changed=2)
       Failed:     3
       -------------
       Total states run:     14
       Total run time:    1.235 s
>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ActionFailed
>>>>>> Message: 1 actions failed.
>>>>>>     Converge failed on instance <default-debian-10-master-py3>.  Please see .kitchen/logs/default-debian-10-master-py3.log for more details
>>>>>> ----------------------
>>>>>> Please see .kitchen/logs/kitchen.log for more details
>>>>>> Also try running `kitchen diagnose --all` for configuration
```
</details>

The `default-debian-10-master-py3` is passing on the `master` branch.

Regards.

tacerus commented:
Hi @baby-gnu, thank you for pointing these out. I repaired the errors. I had to comment out `config_files` from the test pillar, because NSD does not allow conflicting configuration values.

tacerus committed:
feat(zones): allow custom options and files

- allow additional configuration options in zone blocks
- allow installation of zone files from other Salt file roots

Signed-off-by: Georg Pfuetzenreuter <mail@georg-pfuetzenreuter.net>
